src/transformers/trainer.py

```diff
@@ -1701,7 +1701,14 @@ def _prepare_inputs(self, inputs: Dict[str, Union[torch.Tensor, Any]]) -> Dict[s
         """
         for k, v in inputs.items():
             if isinstance(v, torch.Tensor):
-                inputs[k] = v.to(self.args.device)
+                kwargs = dict(device=self.args.device)
+                if self.deepspeed and inputs[k].dtype != torch.int64:
+                    # NLP models inputs are int64 and those get adjusted to the right dtype of the
+                    # embedding. Other models such as wav2vec2's inputs are already float and thus
+                    # may need special handling to match the dtypes of the model
+                    kwargs.update(dict(dtype=self.args.hf_deepspeed_config.dtype()))
+
+                inputs[k] = v.to(**kwargs)
```

Comment:
@sgugger, this one would be of main interest to you.

Reply:
I am not against adding this as a principle, but isn't this a fault of DeepSpeed that should be fixed there? PyTorch AMP does not need us to do this manually.

Reply:
I am not sure my comment communicated the problem. 

- Under NLP models you get inputs of `int` and they get converted to the correct variation of `float` via the embedding amp - so we get fp32 if the model is normal, or fp16 if it's `model.half()` 
- Here the inputs are `float32` - how could deepspeed possibly guess what users inputs are supposed to be? It's the Trainer that needs to set `inputs.to(dtype)`. e.g. how could it know what to do with inputs of `int64`
- AMP doesn't change the dtype of user inputs, since it never works with fp16 directly unless the user set the inputs to it and only adjusts things dynamically to fp16 during some of the computations. 

If you have an idea of how this deriving can be automated, I'm all ears. I'm just not sure how Deepspeed could possibly know what user has in mind.

Do you perhaps propose that if the inputs are fp32 then Deepspeed can automatically convert them to fp16 if it's doing its amp version? But if inputs are anything else than fp32 (e.g. int64`) it should leave them alone?

Reply:
we are also getting ready for new dtypes with the bleed edge pytorch - fp16, tf32, etc. so we should think globally here.

Reply:
> - AMP doesn't change the dtype of user inputs, since it never works with fp16 directly unless the user set the inputs to it and only adjusts things dynamically to fp16 during some of the computations.

That is incorrect. Inside the autocast context manager, the conversion is done automatically, so the user does not have to think about the dtype of the inputs. In the [official AMP tutorial](https://pytorch.org/docs/stable/notes/amp_examples.html#typical-mixed-precision-training) there is no inputs dtype conversion and having int inputs is a specificity of NLP, not the general rule. All CV models have float inputs and I think that is still the majority of Deep Learning models.

DeepSpeed could also properly deal with dtype of the inputs since they wrap the model in their own container. I'm just noticing a point where they could make their API better and more aligned with PyTorch, I have no problem with merging this like I said before.

Reply:
Yes, yes, I'm not arguing, just trying to understand how they could do it in a safe way. We may merge this to move wav2vec2 along and meanwhile I'm talking to Deepspeed about automating this and once it's there to drop my workaround.

> That is incorrect. Inside the autocast context manager, the conversion is done automatically, so the user does not have to think about the dtype of the inputs. In the official AMP tutorial there is no inputs dtype conversion and having int inputs is a specificity of NLP, not the general rule. All CV models have float inputs and I think that is still the majority of Deep Learning models.

I still don't think we are on the same page for AMP. 

I stepped through https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html in debugger and it doesn't covert inputs to fp16 in mixed precision, it sends inputs to Linear in fp32 and converts them on the fly to fp16, returning fp16 result.

That's said: over at Deepspeed we are discussing automatically converting fp32 to fp16 under fp16 enabled and having a special flag to prevent that from happening should someone not want that behavior.

Reply:
I am just talking from a user's perspective. Even if it's inside the CUDA kernel (though I think it's before that ;-) ), AMP does convert the input to FP16 at some point to do the operation, my main point is: the user does not see this and does not have to deal with it. :-)

Reply:
I wrote my comment after I downloaded their example and stepped through with debugger. The input to Linear is fp32 under amp and the conversion happens on C-level.

But I hear you, in the general case 99% of the time it should be safe for deepspeed to do the conversion automatically to match amp/apex.

Thank you for making the suggestion, @sgugger. I made a proposal here: https://github.com/microsoft/DeepSpeed/issues/1144

I was thinking more about the future where we have many more dtypes than just fp16 and fp32.